### ISSUE TITLE
bug report: action_plugins regression in _fixup_perms

### ISSUE THREAD

**Issue by Yannig** [ansible/ansible#17352]
##### COMPONENT NAME

_fixup_perms in **lib/ansible/plugins/action/__init.py**
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel c8431e334b) last updated 2016/09/01 17:49:53 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 7e79c59d38) last updated 2016/09/01 18:01:23 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD e8a5442345) last updated 2016/09/01 18:01:23 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/yannig/dev/ansible-conf/ansible/playbooks/library']
```
##### CONFIGURATION

Nothing special
##### OS / ENVIRONMENT

Nothing special
##### SUMMARY

I have an action plugins that handle tmp file transfert. I use it to upload cert file on remote server and check if there are already in java keytool. The main avantage is that I dont have to worry about the transfert on a tmp location and deletion after my check.

The problem is that since 2.1.2, the method have changed and now I cannot use a path directly. The problem is also present in devel
##### STEPS TO REPRODUCE

Retrieve the playbook + action_plugins + module at the following location: https://github.com/Yannig/yannig-ansible-playbooks/tree/master/fixup_perms

Launch the playbook: `ansible-playbook fixup_perms.yml`
##### EXPECTED RESULTS

```
PLAY [fixup_perms test] ********************************************************

TASK [fixup_perms] *************************************************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   
```
##### ACTUAL RESULTS

``````
PLAY [fixup_perms test] ********************************************************

TASK [fixup_perms] *************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: UnicodeEncodeError: 'ascii' codec can't encode character u'\xe9' in position 46: ordinal not in range(128)
fatal: [localhost]: FAILED! => {"exception": "Traceback (most recent call last):\n  File \"/home/yannig/ansible-2.1.2.0/lib/ansible/executor/task_executor.py\", line 124, in run\n    res = self._execute()\n  File \"/home/yannig/ansible-2.1.2.0/lib/ansible/executor/task_executor.py\", line 450, in _execute\n    result = self._handler.run(task_vars=variables)\n  File \"/home/yannig/dev/yannig/yannig-ansible-playbooks/fixup_perms/action_plugins/fixup_perms.py\", line 21, in run\n    self._fixup_perms(u\"/tmp/resolv.conf\", remote_user)\n  File \"/home/yannig/ansible-2.1.2.0/lib/ansible/plugins/action/__init__.py\", line 365, in _fixup_perms\n    raise AnsibleError('Failed to set file mode on remote files (rc: {0}, err: {1})'.format(res['rc'], res['stderr']))\nUnicodeEncodeError: 'ascii' codec can't encode character u'\\xe9' in position 46: ordinal not in range(128)\n", "failed": true, "msg": "Unexpected failure during module execution.", "stdout": ""}                                                                                        ```
``````

**Comment by Yannig** [comment]
The problem seems to be related to this change: https://github.com/ansible/ansible/pull/16924

I have the following proposition to fix this and be backward compatible:

``` patch
diff --git a/lib/ansible/plugins/action/__init__.py b/lib/ansible/plugins/action/__init__.py
index aecbeca..2b78166 100644
--- a/lib/ansible/plugins/action/__init__.py
+++ b/lib/ansible/plugins/action/__init__.py
@@ -321,6 +321,8 @@ class ActionBase(with_metaclass(ABCMeta, object)):
             # we have a need for it, at which point we'll have to do something different.
             return remote_paths

+        if isinstance(remote_paths, unicode):
+            remote_paths = [ remote_paths ]
         if self._play_context.become and self._play_context.become_user not in ('root', remote_user):
             # Unprivileged user that's different than the ssh user.  Let's get
             # to work!
```

What do you think about?

**Comment by nitzmahone** [comment]
cc @abadger @mattclay

**Comment by mattclay** [comment]
Capturing relevant discussion from IRC:

```
<mattclay> abadger1999: We didn't add backwards compatibility there because the usage has changed. The fixup is no longer recursive.
<abadger1999> mattclay: <nod>  This seems to be a cornercase where the custom action plugin didn't use recursive (it was only uploading a single file).
<mattclay> If we allow a string instead of a list, it should work fine for files. However, if it's a directory it will succeed but not yield correct results.
<abadger1999> :-(
<abadger1999> mattclay: ah -- you mean because of recursive being the default before?
<mattclay> abadger1999: Yes
<abadger1999> yannigperre: Yeah, so this is the safety issue.
<abadger1999> yannigperre: If recursive had been False by default before, we could rely on recursive no longer being a parameter to tell people that they must update their code to set permissions properly.
<abadger1999> yannigperre: But since it was True, people could have been calling fixup_perms with a directory and not specifying the recursive flag.... but the function will no longer recurse the directory so the contained files won't be changed.
<mattclay> abadger1999: We could rename fixup_perms for our internal use, and create a legacy version with the original name. If it's called with recursive False, we could use the new fixup_perms, otherwise raise an error.
<mattclay> abadger1999: But that only fixes the problem if there are users calling it with recursive = False.
<abadger1999> yeah.
<abadger1999> yannigperre: Would that fix your use case?
<abadger1999> yannigperre: sounds like you might have to updatet the action plugin to explicitly pass recurse=False but it would be a single codebase with no conditional once you did that.
<abadger1999> jimi|ansible, mattclay, yannigperre:  If we do the fake _fixupPerms, it would have to go into 2.1.2... Once 2.1.2 is released there will be releases with _fixup_perms in both states.
```

**Comment by mattclay** [comment]
@Yannig See the question @abadger asked in the IRC chat logs I quoted in the previous comment.

**Comment by abadger** [comment]
Hey @Yannig we only have a short window where we can still implement the fake _fixup_perms.  (If 2.1.2 goes out without the fake fixup_perms it will be too late to add it).  Could you let us know ASAP if it will help you to implement that?

**Comment by Yannig** [comment]
Sorry for being late @abadger. I think your proposition is fine.

**Comment by mattclay** [comment]
@Yannig I've added a semi-backwards compatible stub for _fixup_perms that supports non-recursive mode only. It is present in the devel branch and will be included in the next RC for 2.1.2 as well.

Can you confirm that this allows your custom action plugins to work after making sure they set `recursive=False`?

**Comment by mattclay** [comment]
@Yannig The fix for this was included in 2.1.2 RC3 (https://groups.google.com/forum/#!topic/ansible-devel/Ihergewnr6E). Based on your earlier comment that the proposed fix is fine, I'm closing this issue as resolved.

**Comment by Yannig** [comment]
It's fine for me. Just have to add recursive=False in my previous call.
Thanks !
